jzaia_files/grover.py

```diff
@@ -0,0 +1,130 @@
+import numpy as np
+import pennylane as qml
+
+# Define oracles
+ORACLE1_QUBITS   = 6
+ORACLE1_EXPECTED = [1, 1, 0, 1, 0]
+
+ORACLE2_QUBITS   = 10
+ORACLE2_EXPECTED = [1, 0, 1, 0, 1, 0, 1, 0, 1]
+
+ORACLE3_QUBITS   = 17
+ORACLE3_EXPECTED = [0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1]
+
+ORACLES = [(ORACLE1_QUBITS, ORACLE1_EXPECTED),
+           (ORACLE2_QUBITS, ORACLE2_EXPECTED),
+           (ORACLE3_QUBITS, ORACLE3_EXPECTED)]
+
+
+def grovers_setup(num_qubits: int):
+    '''
+    Setup up a circuit for iterations of Grover's search
+
+    Places a circuit into uniform superposition. Additionally, places the
+    ancilla qubit for the oracle in the |-> state such that it can
+    apply phase kickback.
+
+    :param num_qubits: The number of qubits in the circuit
+    '''
+    qml.X(num_qubits-1)
+    for i in range(num_qubits):
+        qml.Hadamard(wires=i)
+
+def grovers_mirror(num_qubits):
+    '''
+    Apply the "Grover's Mirror" reflection to the active circuit
+
+    Performs a reflection across the vector which represents the uniform
+    superposition. This is used for amplitude-amplification for Grover's
+    algorithm.
+
+    :param num_qubits: The number of qubits in the circuit
+    '''
+    for i in range(num_qubits-1):
+        qml.Hadamard(wires=i)
+
+    qml.MultiControlledX(wires=range(num_qubits),
+                         control_values=[False]*(num_qubits-1))
+
+    for i in range(num_qubits-1):
+        qml.Hadamard(wires=i)
+
+def run_grovers(oracle, num_qubits):
+    '''
+    Overall function for running Grover's algorithm on a chosen oracle
+
+    Run Grover's algorithm from start to finish. Prepares a state, and
+    repeats state selection and amplitude-amplification for sqrt(N) iterations
+    (where N = 2^(# of non-ancilla qubits)). This implementation assumes that
+    the oracle always picks precisly 1 state (rather than an arbitrary number).
+
+    :param oracle: A black-box function that acts on a created statevector
+    :param num_qubits: The number of qubits in the circuit the oracle acts
+           on (includes the ancilla)
+    '''
+    grovers_setup(num_qubits)
+
+    reps = int(np.sqrt(2**(num_qubits-1)))
+    for _ in range(reps):
+        oracle()
+        grovers_mirror(num_qubits)
+
+    return [qml.expval(qml.PauliZ(i)) for i in range(num_qubits-1)]
+
+def run_experiment(oracle, num_qubits) -> None:
+    '''
+    Run Grover's algorithm and evaluates results
+
+    Run Grover's algorithm from start to finish, and finds the expected
+    measurement outcome.
+
+    :param oracle: A black-box function that acts on a created statevector
+    :param num_qubits: The number of qubits in the circuit the oracle acts
+           on (includes the ancilla)
+    '''
+    dev = qml.device('lightning.qubit', wires=num_qubits)
+
+
+    circ = qml.QNode(run_grovers, dev)
+
+    expvals = circ(oracle, num_qubits)
+    results = [int(val.numpy() < 0) for val in expvals]
```

Comment:
It shouldn't be. The python virtualenv I'm using does not have tensorflow nor torch installed. The return type I was getting from running the circuit was a list of `pennylane.numpy.tensor.tensor`. All of these tensors are 0-dimensional so I used `.numpy()` to convert these to a scalar value.

Reply:
If I take your python file `grover.py` and run, as it is, in a fresh environment with Python 3.10, where I only installed requirements-dev.txt, I'm getting the following error message:
```python
Traceback (most recent call last):
  File "/home/amintor/Projects/pennylane-lightning/prototypes/jzaia.py", line 114, in <module>
    run_experiment(*gen_oracle(0))
  File "/home/amintor/Projects/pennylane-lightning/prototypes/jzaia.py", line 91, in run_experiment
    results = [int(val.numpy() < 0) for val in expvals]
  File "/home/amintor/Projects/pennylane-lightning/prototypes/jzaia.py", line 91, in <listcomp>
    results = [int(val.numpy() < 0) for val in expvals]
AttributeError: 'float' object has no attribute 'numpy'
```
Would you know why?

Reply:
After you make all sensible updates, could you please re-run your benchmarks in a new and fresh environment?
Please let us know about your results.

Reply:
Okay, I'm seeing this too. It seems the issue was that I originally installed my dependencies from `requirements.txt` instead of `requirements-dev.txt`. It seems there are only a few differences, but namely the dev version installs Pennylane from source, so this is almost certainly related to that. Fixed in 8bd1f93